generator/src/main/java/org/corfudb/generator/Generator.java

```diff
@@ -19,55 +19,196 @@
+     */
+    private void runCpTrimOperation() {
         Runnable cpTrimTask = () -> {
             Operation op = new CheckpointOperation(state);
             op.execute();
         };
-
-        final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
         scheduler.scheduleAtFixedRate(cpTrimTask, 30, cpPeriod, TimeUnit.SECONDS);
+    }
+
+
+    /**
+     * Run a Longevity test that 'freezes' the system every CHUNK_RUNTIME,
+     * sends the correctness log (of one chunk) to verification, then resumes
+     * the run.
+     * The same action happens in a cyclic fashion until the totalRuntime.
+     */
+    public void runLongevityTestForVerification() {
 
-        ExecutorService appWorkers = newWorkStealingPool(numThreads);
+        fillOperationQueue();
+        long lastPollTime = System.currentTimeMillis();
+        consumeOperationQueue();
+        runCpTrimOperation();
 
-        Runnable app = () -> {
-            List<Operation> operations = state.getOperations().sample(numOperations);
-            for (Operation operation : operations) {
-                try {
-                    operation.execute();
-                } catch (Exception ex) {
-                    throw new RuntimeException(ex);
+        while (true) {
+            // One round of 'chunk' starts
+            long cumulativeRuntime = 0;
+            while (true) {
+                Sleep.sleepUninterruptibly(POLL_TIME);
+                // Check if the operationQueue is drained or not,
+                // if it's empty and has not reached CHUNK_RUNTIME,
+                // then fill up the queue for further consuming.
+                if (operationQueue.isEmpty()) {
+                    // Calculate the 'cumulative' runtime so far,
+                    // skipping the time for filling up the operation queue
+                    cumulativeRuntime += System.currentTimeMillis() - lastPollTime;
+                    // Break this while loop if it reaches CHUNK_RUNTIME
+                    if (cumulativeRuntime >= CHUNK_RUNTIME.toMillis()) {
+                        // At the end of the chunk, give some margin
+                        // for the workers to finish their operations
+                        Sleep.sleepUninterruptibly(CHUNK_TIMEOUT);
+                        Correctness.recordOperation("Liveness, " + "Success", false);
+                        break;
+                    }
+                    fillOperationQueue();
+                    lastPollTime = System.currentTimeMillis();
+                } else {
+                    // If it's been more than LIVENESS_TIMEOUT, and the queue is still not drained,
+                    // then the system might have got stuck and we record its 'Liveness' as 'Fail'.
+                    if (System.currentTimeMillis() - lastPollTime > LIVENESS_TIMEOUT.toMillis()) {
+                        Correctness.recordOperation("Liveness, " + "Fail", false);
```

Comment:
Not sure if we need to continue running the remaining operations after we found some chunk already failed? If not, we should probably jump out of the whole loop here.

Reply:
I have discussed this with Anny last Friday. Should we go on running the longevity test if we already found some chunks have correctness issue? And the answer is 'yes': let's allow the test go on even some chunks failed, and it'll be an interesting phenomenon if we found only one/two chunks failed in some middle point and the following chunks all proceed well. We should leave the program running and let human to decide whether to kill a test or not.

Reply:
Yeah make sense to me